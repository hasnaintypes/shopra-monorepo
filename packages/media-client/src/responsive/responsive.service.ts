import { Injectable, Logger } from '@nestjs/common';

import { getMediaProviderClient } from '../imagekit.client';
import type { TransformOptions } from '../types/media.types';
import { mapTransformOptionsToQuery } from '../utils/transformations.util';

@Injectable()
export class ResponsiveService {
  private readonly logger = new Logger(ResponsiveService.name);

  getTransformedUrl(path: string, opts?: TransformOptions): string {
    try {
      const params = mapTransformOptionsToQuery(opts);
      const imageKit = getMediaProviderClient();
      return imageKit.url({ path, transformation: [params] });
    } catch (err) {
      this.logger.error(
        'Failed to generate transformed URL',
        err instanceof Error ? err : new Error(String(err)),
      );
      return '';
    }
  }

  generateSrcset(path: string, widths: number[], opts?: TransformOptions): string {
    try {
      const srcset = widths
        .map((w) => {
          const url = this.getTransformedUrl(path, { ...opts, width: w });
          return `${url} ${w}w`;
        })
        .join(', ');
      return srcset;
    } catch (err) {
      this.logger.error(
        'Failed to generate srcset',
        err instanceof Error ? err : new Error(String(err)),
      );
      return '';
    }
  }
}
