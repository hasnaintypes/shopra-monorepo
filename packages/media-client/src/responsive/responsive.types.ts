export interface SrcsetOptions {
  widths: number[];
  format?: string;
  transforms?: Record<string, string | number>;
}
