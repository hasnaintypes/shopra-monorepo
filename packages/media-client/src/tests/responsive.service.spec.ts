import { ResponsiveService } from '../responsive/responsive.service';

jest.mock('../../src/imagekit.client', () => ({
  getMediaProviderClient: () => ({
    url: ({
      path,
      transformation,
    }: {
      path: string;
      transformation?: Array<Record<string, string | number>>;
    }) => {
      // Simulate ImageKit URL transformation string
      const params =
        transformation && transformation[0]
          ? Object.entries(transformation[0])
              .map(([k, v]) => `${k}-${v}`)
              .join(',')
          : '';
      return `https://fake.imagekit.io${path}?tr:${params}`;
    },
  }),
}));

describe('ResponsiveService', () => {
  let service: ResponsiveService;

  beforeEach(() => {
    service = new ResponsiveService();
  });

  it('getTransformedUrl returns valid URL', () => {
    const url = service.getTransformedUrl('/foo.jpg', { width: 100, format: 'webp' });
    expect(typeof url).toBe('string');
    expect(url).toContain('tr:w-100,f-webp');
  });

  it('generateSrcset returns valid srcset', () => {
    const srcset = service.generateSrcset('/foo.jpg', [320, 640], { format: 'webp' });
    expect(typeof srcset).toBe('string');
    expect(srcset).toContain('320w');
    expect(srcset).toContain('640w');
  });

  it('generateSrcset snapshot', () => {
    const srcset = service.generateSrcset('/foo.jpg', [320, 640, 1280], { format: 'webp' });
    expect(srcset).toMatchSnapshot();
  });

  // TODO: test logger on error
});
