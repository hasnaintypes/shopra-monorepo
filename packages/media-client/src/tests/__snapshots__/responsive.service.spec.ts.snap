// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ResponsiveService generateSrcset snapshot 1`] = `"https://fake.imagekit.io/foo.jpg?tr:w-320,f-webp 320w, https://fake.imagekit.io/foo.jpg?tr:w-640,f-webp 640w, https://fake.imagekit.io/foo.jpg?tr:w-1280,f-webp 1280w"`;
